mathjax3-ts/ui/menu/MJContextMenu.ts

```diff
@@ -0,0 +1,249 @@
+        if (!menu) {
+            ContextMenu.MenuUtil.error(null, 'Wrong JSON format for menu.');
+            return;
+        }
+        const {pool, items, id} = menu;
+        const ctxtMenu = new this();
+        const menuPool = ctxtMenu.getPool();
+        pool.forEach(({name, getter, setter}: PoolItem) => {
+            menuPool.insert(new ContextMenu.Variable(name, getter, setter));
+        });
+        ctxtMenu.parseItems(items);
+        return ctxtMenu;
+    }
+
+    /*======================================================================*/
+
+    /**
+     * The MathItem that has posted the menu
+     */
+    public mathItem: MathItem<HTMLElement, Text, Document> = null;
+
+    /**
+     * The annotation selected in the Annotation submenu (neede for the info box to be able to show it)
+     */
+    public annotation: string = '';
+
+    /**
+     * The info box for showing annotations (created by the Menu object that contains this MJContextMenu)
+     */
+    public showAnnotation: SelectableInfo;
+
+    /**
+     * The function to copy the selected annotation (set by the containing Menu item)
+     */
+    public copyAnnotation: () => void;
+
+    /**
+     * The annotation types to look for in a MathItem
+     */
+    public annotationTypes: {[type: string]: string[]} = {};
+
+    /*======================================================================*/
+
+    /**
+     * Before posting the menu, set the name for the ShowAs and CopyToClipboard menus,
+     *   enable/disable the semantics check item, and get the annotations for the MathItem
+     *
+     * @override
+     */
+    public post(x?:any, y?: number) {
+        if (this.mathItem) {
+            if (y !== undefined) {
+                // FIXME:  handle error output jax
+                const input = this.mathItem.inputJax.name;
+                const original = this.findID('Show', 'Original');
+                original.content = (input === 'MathML' ? 'Original MathML' : input + ' Commands');
+                const clipboard = this.findID('Copy', 'Original');
+                clipboard.content = original.content;
+                const semantics = this.findID('Settings', 'semantics');
+                input === 'MathML' ? semantics.disable() : semantics.enable();
+                this.getAnnotationMenu();
+            }
+            super.post(x, y);
+        }
+    }
+
+    /**
+     * Clear the stored MathItem when the menu is removed
+     *
+     * @override
+     */
+    public unpost() {
+        super.unpost();
+        this.mathItem = null;
+    }
+
+    /*======================================================================*/
+
+    /**
+     * Find an item in the menu (recursively descending into submenus, if needed)
+     *
+     * @param {string[]} names   The menu IDs to look for
+     * @returns {string}         The menu item (or null if not found)
+     */
+    public findID(...names: string[]) {
```

Comment:
Not quite.  Because the IDs are actually the localization IDs, they are not necessarily unique, and so the `find()` you indicate will find the first item with a given ID.  This function explicitly gives a submenu path (by naming each submenu along the way individually).  For example, in the Explorer color menus, all the colors with the same names have the same IDs, and the entries in the "Save Math As" and "Copy to Clipboard" menus have the same ids (because their localization phrases are the same).